java/arcs/android/util/ProtoPrefetcher.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.android.util
+
+import android.os.Trace
+import arcs.android.crdt.CrdtCountProto
+import arcs.android.crdt.CrdtDataProto
+import arcs.android.crdt.CrdtEntityProto
+import arcs.android.crdt.CrdtEntityReferenceProto
+import arcs.android.crdt.CrdtOperationProto
+import arcs.android.crdt.CrdtSetProto
+import arcs.android.crdt.CrdtSingletonProto
+import arcs.android.crdt.RawEntityProto
+import arcs.android.crdt.ReferencablePrimitiveProto
+import arcs.android.crdt.ReferencableProto
+import arcs.android.crdt.ReferencableSetProto
+import arcs.android.crdt.ReferenceProto
+import arcs.android.crdt.VersionMapProto
+import arcs.android.storage.ProxyMessageProto
+import java.util.concurrent.Executors
+
+/**
+ * Run asynchronously as a cache warmer to suffer from Arcs protos initialization
+ * overhead as well as cache misses on behalf of whoever needs Arcs capabilities.
+ *
+ * The initialization overhead typically comes from instantiating default instance
+ * and parsing/building proto schema.
+ *
+ * Recommend executing [ProtoPrefetcher.prefetch] as early as possible i.e.
+ * the entry of Application lifecycle: `onCreate()`
+ */
+object ProtoPrefetcher {
+    /** Identify [ProtoPrefetcher] in thread lists and system traces. */
+    private const val name = "ProtoPrefetcher"
+
+    /**
+     * Single-threaded executor.
+     *
+     * TODO(ianchang): tune background priority if needed.
+     */
+    private val executor = Executors.newFixedThreadPool(1) {
+        Thread(it).apply { name = this@ProtoPrefetcher.name }
+    }
+
+    /** Procedures to be taken to warm up Arcs protos. */
+    private val procedures = arrayOf<() -> Any>(
+        ProxyMessageProto::getDefaultInstance,
+        ProxyMessageProto.newBuilder()::build,
+        ProxyMessageProto.SyncRequest::getDefaultInstance,
+        ProxyMessageProto.SyncRequest.newBuilder()::build,
```

Comment:
I don't think you need both lines since newBuilder() uses getDefaultInstance
https://cs.corp.google.com/piper///depot/google3/blaze-out/genfiles/third_party/java_src/arcs/java/arcs/android/storage/proxy_message_proto-lite-src.jar.files/arcs/android/storage/ProxyMessageProto.java?rcl=308940462&l=1244
maybe try verify that the newBuilder is enough

Reply:
From generated code, yes.
But from the reality, no.

When I worked on the list yesterday, found using the same list introduced different overhead on client and server (iirc, roughly 1: 4).
I don't have a correct explanation on this but it looks like more parts of the list are ignored at client as if there is more dead code being eliminated.

So the best way is to maintain a full list.
i.e., one might only use getDefaultInstance but never builds the proto.
// TODO: not implemented yet, just return default values
fun foo() = proto.getDefaultInstance()